packages/vats/src/ibc.js

```diff
@@ -61,6 +62,180 @@ const prepareAckWatcher = zone => {
   return makeAckWatcher;
 };
 
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ * @param {ReturnType<import('@agoric/vow').prepareVowTools>} powers
+ * @param {ReturnType<prepareProtocolImplAttemptWatcher>} makeProtocolImplAttemptWatcher
+ */
+const prepareProtocolImplInboundWatcher = (
+  zone,
+  { watch },
+  makeProtocolImplAttemptWatcher,
+) => {
+  const makeProtocolImplInboundWatcher = zone.exoClass(
+    'ProtocolImplInboundWatcher',
+    M.interface('ProtocolImplInboundWatcher', {
+      onFulfilled: M.call(M.any()).rest(M.any()).returns(M.any()),
+    }),
+    ({
+      channelKeyToAttempt,
+      channelKeyToInfo,
+      obj,
+      util,
+      rPortID,
+      rChannelID,
+      hops,
+    }) => ({
+      channelKeyToAttempt,
+      channelKeyToInfo,
+      obj,
+      util,
+      rPortID,
+      rChannelID,
+      hops,
+    }),
+    {
+      onFulfilled(attempt) {
+        const {
+          channelKeyToAttempt,
+          channelKeyToInfo,
+          obj,
+          util,
+          rPortID,
+          rChannelID,
+          hops,
+        } = this.state;
+
+        const { channelID, portID, asyncVersions, order, version } = obj;
+
+        // Tell what version string we negotiated.
+        return watch(
+          E(attempt).getLocalAddress(),
+          makeProtocolImplAttemptWatcher({
+            channelID,
+            portID,
+            channelKeyToAttempt,
+            channelKeyToInfo,
+            attempt,
+            obj,
+            util,
+            asyncVersions,
+            rPortID,
+            rChannelID,
+            order,
+            hops,
+            version,
+          }),
+        );
+      },
+    },
+  );
+  return makeProtocolImplInboundWatcher;
+};
+
+/** @param {import('@agoric/base-zone').Zone} zone */
+const prepareProtocolImplAttemptWatcher = zone => {
+  const makeProtocolImplAttemptWatcher = zone.exoClass(
```

Comment:
If we rely on `watch`'s propagating its third argument, `watchContext`, we only need a singleton watcher.
```suggestion
const makeProtocolImplAttemptWatcher = zone => {
  const protocolImplAttemptWatcher = zone.exo(
```